Постановка задачи задачи:


$$\min_{D} ((-(1.5(x_1 + 0.5)^2) + x_2 + 2)^2 + x_2^2) = ?$$


$$D = \{x\in\R^2| g_1(x) \leqslant 0, g_2(x)\leqslant 0, g_3(x) \leqslant 0\}$$
$$g_1(x) = -0.2x_1^2 - x_2 + 0.5,
g_2(x) = -x_1 - x_2 - 1,
g_3(x) = x_1^2 - 3x_1 + 2.25 - (x_2^2 - 3x_2 + 2.25) + 2$$


In [1]:
import sympy as sp
from sympy import pprint, latex
from IPython.display import display, Latex
from itertools import chain, combinations

def powerset(iterable):
    s = list(iterable)
    return chain.from_iterable(combinations(s, r) for r in range(len(s)+1))

In [2]:
X = sp.symbols(['x_1', 'x_2'], real=True)
lamds = sp.symbols([rf'\lambda_{i}' for i in range(1, 4)], real=True)

F = (-(1.5 * (X[0] + 0.5)**2) + X[1] + 2)**2 + X[1]**2
D = [-0.2*X[0]**2 - X[1] + 1/2, -X[0] - X[1] - 1, X[0]**2 - 3 * X[0] + 2.25 - (X[1]**2 - 3 * X[1] + 2.25) + 2]

L = F
for i, d in enumerate(D):
    L = L + d * lamds[i]

print("Лагранжиан:")
display(Latex, L)
dL = [sp.diff(L, X[0]), sp.diff(L, X[1])]
print("Градиент Лагранжиана:")
display(Latex, sp.Matrix(dL))

Лагранжиан:


IPython.core.display.Latex

\lambda_1*(-0.2*x_1**2 - x_2 + 0.5) + \lambda_2*(-x_1 - x_2 - 1) + \lambda_3*(x_1**2 - 3*x_1 - x_2**2 + 3*x_2 + 2) + x_2**2 + 4*(x_2/2 - 0.75*(x_1 + 0.5)**2 + 1)**2

Градиент Лагранжиана:


IPython.core.display.Latex

Matrix([
[-0.4*\lambda_1*x_1 - \lambda_2 + \lambda_3*(2*x_1 - 3) + 4*(-3.0*x_1 - 1.5)*(x_2/2 - 0.75*(x_1 + 0.5)**2 + 1)],
[                              -\lambda_1 - \lambda_2 + \lambda_3*(3 - 2*x_2) + 4*x_2 - 3.0*(x_1 + 0.5)**2 + 4]])

In [3]:
sols = []

In [4]:
cases = powerset([0, 1, 2])

for c in cases:
    sys = sp.Matrix([dL[0], dL[1]] + [D[i] for i in c] + [lamds[j] for j in [0, 1, 2] if j not in c])
    sols.append(sp.solve(sys, X + lamds))
print(sols)

[[(-1.65470053837925, 0.0, 0.0, 0.0, 0.0), (-0.500000000000000, -1.00000000000000, 0.0, 0.0, 0.0), (0.654700538379252, 0.0, 0.0, 0.0, 0.0)], [], [], [], [(-1.20809924354783, 0.208099243547831, -1.92668757447740, 5.25487093252969, 0.0), (6.20809924354783, -7.20809924354783, 2080.67668757448, -2240.50487093253, 0.0)], [(-5.57406614872846, -5.71404268608011, 1107.29202247182, 0.0, 83.4058438123157), (1.44123873066190, 0.0845661842480148, 66.7777659646278, 0.0, 26.0502200227294)], [(-0.250000000000000, -0.750000000000000, 0.0, -0.620117187500000, -0.318359375000000)], []]


In [5]:
true_sols = []
for M in sols:
    if (M == []):
        continue
    for s in M:
        ok = True
        if (D[i].subs(zip(X + lamds, s)).evalf() > 0):
            ok = False
        for l in s[2:]:
            if (l < 0):
                ok = False
        if (ok):
            true_sols.append(s)

print(true_sols)
func = sp.lambdify(X, F)
func_values = [func(s[0], s[1]) for s in true_sols if s != []]
print(func_values)

min_idx = min(range(len(func_values)), key= lambda i: func_values[i])
print(f"Min value: {func_values[min_idx]} in x = ({true_sols[min_idx][0]}, {true_sols[min_idx][1]})")

[(-0.500000000000000, -1.00000000000000, 0.0, 0.0, 0.0), (-5.57406614872846, -5.71404268608011, 1107.29202247182, 0.0, 83.4058438123157)]
[2.00000000000000, 1824.75549157257]
Min value: 2.00000000000000 in x = (-0.500000000000000, -1.00000000000000)


Получаем, что минимум исследуемой функции достигается в точке $(-0.500000000000000, -1.00000000000000)$ и равен $2$